zkevm-circuits/src/evm_circuit.rs

```diff
@@ -1 +1,930 @@
+            Self::AccountStorage { .. } => F::from_u64(Target::Storage as u64),
+            _ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Eq, Ord)]
+pub(crate) enum FixedLookup {
+    // Noop provides [0, 0, 0, 0] row
+    Noop,
+    // meaningful tags start with 1
+    Range256,
+    Range32,
+    BitwiseAnd,
+    BitwiseOr,
+    BitwiseXor,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum Lookup<F> {
+    FixedLookup(FixedLookup, [Expression<F>; 3]),
+    BusMappingLookup(BusMappingLookup<F>),
+}
+
+#[derive(Clone, Debug)]
+struct Constraint<F> {
+    name: &'static str,
+    selector: Expression<F>,
+    polys: Vec<Expression<F>>,
+    lookups: Vec<Lookup<F>>,
+}
+#[derive(Clone, Debug)]
+pub(crate) struct Cell<F> {
+    // expression for constraint
+    expression: Expression<F>,
+    // relative position to selector for synthesis
+    column: Column<Advice>,
+    rotation: usize,
+}
+
+impl<F: FieldExt> Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        value: Option<F>,
+    ) -> Result<circuit::Cell, Error> {
+        region.assign_advice(
+            || {
+                format!(
+                    "Cell column: {:?} and rotation: {}",
+                    self.column, self.rotation
+                )
+            },
+            self.column,
+            offset + self.rotation,
+            || value.ok_or(Error::SynthesisError),
+        )
+    }
+}
+
+// TODO: Integrate with evm_word
+#[derive(Clone, Debug)]
+struct Word<F> {
+    // random linear combination expression of cells
+    expression: Expression<F>,
+    // inner cells for syntesis
+    cells: [Cell<F>; 32],
+}
+
+impl<F: FieldExt> Word<F> {
+    fn new(cells: &[Cell<F>], r: F) -> Self {
+        let r = Expression::Constant(r);
+        Self {
+            expression: cells
+                .iter()
+                .rev()
+                .fold(Expression::Constant(F::zero()), |acc, byte| {
+                    acc * r.clone() + byte.expr()
+                }),
+            cells: cells.to_owned().try_into().unwrap(),
+        }
+    }
+
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        word: Option<[u8; 32]>,
+    ) -> Result<Vec<circuit::Cell>, Error> {
+        word.map_or(Err(Error::SynthesisError), |word| {
+            self.cells
+                .iter()
+                .zip(word.iter())
+                .map(|(cell, byte)| {
+                    cell.assign(region, offset, Some(F::from_u64(*byte as u64)))
+                })
+                .collect()
+        })
+    }
+}
+
+// TODO: Should we maintain these state in circuit or we prepare all of them
+// when parsing debug trace?
+#[derive(Debug)]
+pub(crate) struct CoreStateInstance {
```

Comment:
I would prepare that during the debug step. So that reduces the complexity from this crate. And we have much more data avaliable in the trace than we have here when the trace has already been parsed.

Reply:
Yeah make sense, this is only for making the current work testable.

Reply:
Perfect. Should we file an issue for that?

Reply:
Mentioned in #61